software/controller/lib/hal/gpio.h

```diff
@@ -24,6 +24,9 @@ Reference abbreviations [RM], [DS], etc are defined in hal/README.md.
 
 #include <cstdint>
 
+#include "adc.h"
```

Comment:
I'd like to question if functionally this dependency is really going in the  correct direction or only for the sake of convenience of interface? Just the impression I get, but maybe I am wrong?